sip/lib/sip_endpoint.ex

```diff
@@ -0,0 +1,731 @@
+  def handle_info({:call_info, {:call_ready, _opts}}, _ctx, %{endpoint_state: :in_call} = state) do
+    Logger.warning(
+      "SIP Endpoint: Received `:call_ready` info, but the pipelines are already spawned. Ignoring"
+    )
+
+    {[], state}
+  end
+
+  @impl true
+  def handle_info({:call_info, {:end, reason} = msg}, _ctx, state) do
+    case reason do
+      :cancelled ->
+        Logger.info("SIP Endpoint: Call attempt cancelled by user")
+
+      :user_hangup ->
+        Logger.info("SIP Endpoint: Call ended by user")
+
+      :busy ->
+        Logger.info("SIP Endpoint: Call declined, other side is busy")
+
+      :declined ->
+        Logger.info("SIP Endpoint: Call declined by other side")
+
+      :normal_clearing ->
+        Logger.info("SIP Endpoint: Call ended by other side (hangup)")
+
+      reason ->
+        Logger.warning("SIP Endpoint: Call ended with reason: #{inspect(reason)}")
+    end
+
+    actions = [
+      notify_parent: {:forward_to_parent, msg},
+      notify_parent: :finished
+    ]
+
+    actions =
+      if state.endpoint_state != :in_call do
+        [notify_parent: :ready] ++ actions
+      else
+        actions
+      end
+
+    {actions, %{state | endpoint_state: :terminating}}
+  end
+
+  @impl true
+  def handle_info(info, _ctx, state) do
+    Logger.warning("SIP Endpoint: Unexpected info: #{inspect(info)}. Ignoring.")
+    {[], state}
+  end
+
+  @impl true
+  def handle_terminate_request(_ctx, state) do
+    Logger.debug("SIP Endpoint: Received terminate request")
+
+    case state.endpoint_state do
+      :calling ->
+        OutgoingCall.cancel(state.call_id)
+
+      :in_call ->
+        OutgoingCall.bye(state.call_id)
+
+      _other ->
+        nil
+    end
+
+    Call.stop(state.register_call_id)
+
+    {[terminate: :normal], %{state | endpoint_state: :terminating}}
+  end
+
+  defp try_calling(state, playback_state, phone_number) do
+    case state.endpoint_state do
+      _any when playback_state != :playing ->
+        Logger.info("SIP Endpoint: Postponing call until in state playing")
+        %{state | phone_number: phone_number}
+
+      :unregistered ->
+        Logger.info("SIP Endpoint: Postponing call until registered")
+        %{state | phone_number: phone_number, endpoint_state: :unregistered_call_pending}
+
+      :registered ->
+        Logger.info("SIP Endpoint: Calling #{inspect(phone_number)}...")
+        state = %{state | phone_number: phone_number}
+        {call_id, _pid} = spawn_call(state)
+        %{state | call_id: call_id, endpoint_state: :calling}
+
+      _other ->
+        Logger.warning("SIP Endpoint: Already calling, or endpoint is terminating")
+        state
+    end
+  end
+
+  defp spawn_call(state, module \\ OutgoingCall) do
+    state
+    |> case do
+      %State{} = struct -> Map.from_struct(struct)
+      %{} = map -> map
+      _other -> raise "State is not map nor struct"
+    end
+    |> Map.put(:endpoint, self())
+    |> then(&struct(Call.Settings, &1))
```

Comment:
Maybe the module `Call.Settings` should have a function `new`, where we would provide as arguments: `State` and `endpoint_pid`?